staging/src/k8s.io/apiserver/pkg/storage/cacher/cacher.go

```diff
@@ -667,19 +728,41 @@ func (c *Cacher) processEvent(event *watchCacheEvent) {
 }
 
 func (c *Cacher) dispatchEvents() {
+	startID := bucketID(c.clock.Now())
```

Comment:
This logic means that, we can potentiall send a bookmark even almost 3s before timeout:
- if your timeout finishes at say Xs99999ms that will hit bucket X-2
- if the time now is (X-2)s00000ms, that is bucket X-2
- so we can send bookmark ~3s before timeout

I'm actually wondering what are the exact benefits from bucketing here? What if instead we would be keeping a simple list sorted by timeout? I'm not saying that we should for sure change it, but i would like to understand the exact trade-off here.

Reply:
It's simpler and more efficient to use buckets, and I do not think it requires very high accuracy here.
If we find that the accuracy is important, we may need to trigger bookmark by the timer in watch.go (which is non trivial now).

Reply:
> which is non trivial now
It will be if we refactor the timer to watcher.

Reply:
I agree we don't need very high accuracy here.
But if we're explicitly making this assumption, this at the very least requires an explicit comment what the semantic will be.